Add message details for AmbiguousMatchException

IDisposable opened the pull request:
Add information to what is ambiguously matched when an AmbiguousMatchException or an AmbiguousInvovationException is thrown as we know what we found multiple matching items for.

Fixes #18255

Basic idea is to use the known-first match as a source for the declaring type name and member name to format into the Exception message value.

There's a bunch of similar/common/duplicated code all over, so not sure where (which subset) to apply so this PR attempts most of the uses of `AmbigousMatchException` and `AmbiguousInvocationException` (which has similar issues).

Need to know before this gets final which of the libraries/builds this needs to be done in:
- [x] src/coreclr/nativeaot/System.Private.CoreLib
- [x] src/libraries/System.Private.CoreLib
- [x] src/libraries/System.Reflection.MetadataLoadContext
- [x] src/mono/System.Private.CoreLib

I also noted that the same localizable string ends up in several assemblies the way things are now (e.g. there's a Strings.resx and matching ThrowHelper.cs for each of the target libraries above, with some overlapped strings).  I am not sure there is anything to be done about that, but worth calling out in this PR

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:
```diff
@@ -2878,10 +2876,10 @@ public override InterfaceMapping GetInterfaceMap([DynamicallyAccessedMembers(Dyn
             if (types == null || types.Length == 0)
             {
                 // no arguments
+                PropertyInfo firstCandidate = candidates[0];
```
Hoisted this assignment up so we can use in the throw below.

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:
```diff
@@ -2891,7 +2889,7 @@ public override InterfaceMapping GetInterfaceMap([DynamicallyAccessedMembers(Dyn
                     if (returnType is null)
                         // if we are here we have no args or property type to select over and we have more than one property with that name
-                        throw new AmbiguousMatchException();
+                        ThrowHelper.ThrowAmbiguousMatchException(firstCandidate);
```
As this is the first use of the helper in this PR,  I thought I would mention the strategy. To resolve the original issue and add more information to figure out _what_ was ambiguous, we need to know the name of the member that has ambiguity. 

Since `PropertyInfo` et al all descend from `MemberInfo`, we can make a helper that formats messages from calling `.ToString()` on the info and it's `DeclaringType`. This yields, for example from `String.Length` the values `System.String` and `Int32 Length` as the insertions. See _ThrowHelper.cs_

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:
```diff
@@ -2920,7 +2918,7 @@ public override InterfaceMapping GetInterfaceMap([DynamicallyAccessedMembers(Dyn
                 {
                     if (match != null)
-                        throw new AmbiguousMatchException();
+                        ThrowHelper.ThrowAmbiguousMatchException(match);
```
We have the previously matched one in `match` and we've encountered another one, so we can use `match` for the info.

IDisposable commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/BindingFlagSupport/QueryResult.cs:
```diff
@@ -112,11 +112,10 @@ public void CopyTo(MemberInfo[] array, int startIndex)
-
-                        if (!_policies.OkToIgnoreAmbiguity(match, challenger))
-                            throw new AmbiguousMatchException();
+                        // If they're not the same method, we throw if the policy doesn't allow ambiguity.
```
Merged the two `if` statements as they are logically doing the full check, and in either case throwing with the same info.

buyaa-n replied:
It doesn't seem like for method only, in general the 1st part of the sentence doesn't make sense to me.
```suggestion
                        // If they're not from same type, we throw if the policy doesn't allow ambiguity.
```

IDisposable commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/TypeInfos/RuntimeTypeInfo.BindingFlags.cs:
```diff
@@ -155,18 +155,21 @@ protected sealed override PropertyInfo GetPropertyImpl(string name, BindingFlags
                 if (types == null || types.Length == 0)
                 {
                     // no arguments
+                    PropertyInfo firstCandidate = candidates[0];
```
Hoisted this up to use in the throw.

IDisposable commented on src/coreclr/nativeaot/System.Private.TypeLoader/src/Internal/Reflection/Execution/AssemblyBinderImplementation.cs:
```diff
@@ -82,7 +82,7 @@ public sealed override bool Bind(RuntimeAssemblyName refName, bool cacheMissedLo
                     if (foundMatch)
                     {
-                        exception = new AmbiguousMatchException();
+                        exception = new AmbiguousMatchException(SR.Format(SR.AmbiguousMatchException_Assembly, refName.FullName));
```
This is _similar_ to the other cases, but in the path we can directly talk about the `refName.FullName` which has everything we need to disambiguate.

IDisposable commented on src/coreclr/nativeaot/System.Private.TypeLoader/src/Resources/Strings.resx:
```diff
@@ -1,17 +1,17 @@
-  <!-- 
-    Microsoft ResX Schema 
-    
+  <!--
```
Forgive the autosave reformatting, I can revert the white-space changes if it matters.

buyaa-n replied:
NIT: normally we revert unrelated changes

IDisposable commented on src/coreclr/nativeaot/System.Private.TypeLoader/src/Resources/Strings.resx:
```diff
@@ -129,4 +129,7 @@
   <data name="FileLoadException_RefDefMismatch" xml:space="preserve">
     <value>Cannot load assembly '{0}'. The assembly exists but its version {1} is lower than the requested version {2}.</value>
   </data>
+  <data name="AmbiguousMatchException_Assembly" xml:space="preserve">
```
Used only by `AssemblyBinderImplementation`

IDisposable commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -175,15 +175,27 @@
   <data name="AmbiguousImplementationException_NullMessage" xml:space="preserve">
     <value>Ambiguous implementation found.</value>
   </data>
+  <data name="AmbiguousImplementationException_Simple" xml:space="preserve">
```
Used when we _do not_ have a `DeclaringType` to `.ToString()`.

IDisposable commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -175,15 +175,27 @@
+  <data name="AmbiguousImplementationException_Simple" xml:space="preserve">
+    <value>Ambiguous implementation found for: '{0}'.</value>
+  </data>
+  <data name="AmbiguousImplementationException_Detailed" xml:space="preserve">
```
Used when we _do_ have a `DeclaringType` to `.ToString()` .

IDisposable commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -175,15 +175,27 @@
     <value>Attempted to read or write protected memory. This is often an indication that other memory is corrupt.</value>
   </data>
-  <data name="Arg_AmbiguousMatchException" xml:space="preserve">
+  <data name="Arg_AmbiguousMatchException_NullMessage" xml:space="preserve">
```
Renamed this one to match the pattern used by the resx messages for `AmbiguousImplementationExeception_NullMessage`

buyaa-n replied:
NIT: looks better to be named `Arg_AmbiguousMatchException_NoMessage`, feel free to update the AmbiguousImplementationExeception_NullMessage` too
```suggestion
  <data name="Arg_AmbiguousMatchException_NoMessage" xml:space="preserve">
```

IDisposable commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -175,15 +175,27 @@
+  <data name="Arg_AmbiguousMatchException_NullMessage" xml:space="preserve">
     <value>Ambiguous match found.</value>
   </data>
+  <data name="Arg_AmbiguousMatchException_Simple" xml:space="preserve">
```
w/o `DeclaringType`

IDisposable commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -175,15 +175,27 @@
+  <data name="Arg_AmbiguousMatchException_Simple" xml:space="preserve">
+    <value>Ambiguous match found for: '{0}'.</value>
+  </data>
+  <data name="Arg_AmbiguousMatchException_Detailed" xml:space="preserve">
```
w/ `DeclaringType`

buyaa-n replied:
Not sure if we really need this

IDisposable commented on src/libraries/System.Private.CoreLib/src/System/DefaultBinder.cs:
```diff
@@ -383,8 +383,10 @@ internal partial class DefaultBinder : Binder
 #endregion
             }
 
+            MethodBase bestMatch = candidates[currentMin]!;
```
Hoisted up to use in the throw (note the trailing `!` that was in all the references below. We KNOW it's no-null.

IDisposable commented on src/libraries/System.Private.CoreLib/src/System/DefaultBinder.cs:
```diff
@@ -395,7 +397,7 @@ internal partial class DefaultBinder : Binder
             // If the parameters and the args are not the same length or there is a paramArray
             //  then we need to create a argument array.
-            ParameterInfo[] parameters = candidates[currentMin]!.GetParametersNoCopy();
+            ParameterInfo[] parameters = bestMatch.GetParametersNoCopy();
```
No reason to keep derefing the array, we know what it is above (miniscule performance increase _possible_).

IDisposable commented on src/libraries/System.Private.CoreLib/src/System/DefaultBinder.cs:
```diff
@@ -526,9 +528,10 @@ public sealed override FieldInfo BindToField(BindingFlags bindingAttr, FieldInfo
                     }
                 }
             }
+            FieldInfo bestMatch = candidates[currentMin];
```
Hoisted out of the return to reuse in the throw

IDisposable commented on src/libraries/System.Private.CoreLib/src/System/DefaultBinder.cs:
```diff
@@ -620,9 +623,10 @@ public sealed override FieldInfo BindToField(BindingFlags bindingAttr, FieldInfo
                     }
                 }
             }
+            MethodBase bestMatch = candidates[currentMin];
```
Hoisted out of the return to reuse in the throw

IDisposable commented on src/libraries/System.Private.CoreLib/src/System/Reflection/AmbiguousMatchException.cs:
```diff
@@ -10,7 +10,7 @@ namespace System.Reflection
     {
         public AmbiguousMatchException()
-            : base(SR.Arg_AmbiguousMatchException)
+            : base(SR.Arg_AmbiguousMatchException_NullMessage)
```
Renamed in the resx to match the AmbigiousInvocationException_NullMessage` pattern.

IDisposable commented on src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs:
```diff
@@ -575,6 +577,34 @@ internal static void ThrowFormatIndexOutOfRange()
+           throw GetAmbiguousMatchException(memberInfo);
+        }
+
+        internal static AmbiguousMatchException GetAmbiguousMatchException(MemberInfo memberInfo)
```
Made this an internal helper  because I originally thought it might be useful when we wanted an Exception to pass out (not throw), but it turns out there's only one place we do that, and it's not here... should I merge this into the method above?

Note: I copied the pattern for `GetArraySegmentCtorValidationFailedException` below, so there is precedent for this construction in this class...

buyaa-n replied:
If it is not used for passing an exception probably throwing here same as above better, in case you update this method do not forget to add the `[DoesNotReturn]` attribute.

IDisposable commented on src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs:
```diff
@@ -575,6 +577,34 @@ internal static void ThrowFormatIndexOutOfRange()
+        {
+            var declaringType = memberInfo.DeclaringType;
+            return (declaringType is not null)
+                ? new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Detailed, declaringType.ToString(), memberInfo.ToString()))
```
Using the example of being passed the `MethodInfo` of  the property`System.String.Length`....

So if we **do have** a non-null `.DeclaringType`, we can use that to format up with more detail (yielding for example "System.String" for the first part.

If we **do not**, then we just emit the `memberInfo.ToString()` (which yields for example "Int32 Length")

buyaa-n replied:
Not sure if adding the `DeclaringType` info that useful ...

IDisposable commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/CustomAttributes/CustomAttributeHelpers.cs:
```diff
@@ -18,7 +18,7 @@ public static CustomAttributeNamedArgument ToCustomAttributeNamedArgument(this T
                 throw new MissingMemberException(attributeType.FullName, name);
             if (members.Length > 1)
-                throw new AmbiguousMatchException();
+                throw ThrowHelper.GetAmbiguousMatchException(members[0]);
```
Maybe I should hoist this one to follow pattern better?

buyaa-n replied:
better to follow the same pattern

IDisposable commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/RuntimeTypeInfo.BindingFlags.cs:
```diff
@@ -136,10 +136,11 @@ internal partial class RoType
                 // For perf and .NET Framework compat, fast-path these specific checks before calling on the binder to break ties.
                 if (types == null || types.Length == 0)
                 {
+                    PropertyInfo firstCandidate = candidates[0];
```
Hoisted to use with the throw.

IDisposable commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/General/ThunkedApis.cs:
```diff
@@ -199,7 +199,7 @@ public sealed override Type[] GetGenericArguments()
                     continue;
                 if (match != null)
-                    throw new AmbiguousMatchException();
+                    ThrowHelper.ThrowAmbiguousMatchException(match);
```
Again, when we hit this line, `match` is the _first_ one we found, and we're going to throw with it's info

IDisposable commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/Runtime/BindingFlagSupport/QueryResult.cs:
```diff
@@ -112,11 +112,11 @@ public void CopyTo(MemberInfo[] array, int startIndex)
                         // that means we let the first match win - unless, of course, they're both the "most derived member".
                         if (match.DeclaringType!.Equals(challenger.DeclaringType))
-                            throw new AmbiguousMatchException();
+                            throw ThrowHelper.GetAmbiguousMatchException(match);
```
Could merge these lines like we do in `QueryResult.cs`. Thoughts as to which pattern looks better?

IDisposable commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Types/Ecma/EcmaDefinitionType.BindingFlags.cs:
```diff
@@ -94,7 +94,7 @@ internal sealed override IEnumerable<RoType> GetNestedTypesCore(NameFilter? filt
                 {
                     if (match != null)
-                        throw new AmbiguousMatchException();
+                        throw ThrowHelper.GetAmbiguousMatchException(match);
```
Once again, `match` is referring to the first one.

IDisposable commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Types/RoType.GetInterface.cs:
```diff
@@ -33,7 +33,7 @@ internal partial class RoType
                     continue;
                 if (match != null)
-                    throw new AmbiguousMatchException();
+                    throw ThrowHelper.GetAmbiguousMatchException(match);
```
... and again.

IDisposable commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/ThrowHelper.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    internal static class ThrowHelper
+    {
+        internal static AmbiguousMatchException GetAmbiguousMatchException(MemberInfo memberInfo)
```
Kinda hating that we have the same helpers and strings in multiple places, but I don't know/think this is worth a public API change or making internals visible to others...

IDisposable commented on src/mono/System.Private.CoreLib/src/System/Reflection/Emit/RuntimeTypeBuilder.Mono.cs:
```diff
@@ -430,7 +430,7 @@ protected override void AddInterfaceImplementationCore([DynamicallyAccessedMembe
                 {
                     if (count > 1)
-                        throw new AmbiguousMatchException();
+                        ThrowHelper.ThrowAmbiguousMatchException(found!);
```
`found` is known non null here, but the compiler doesn't grok that. It comes from the `foreach (` over `cb`  so the `!` is warranted.

Could do a tiny bit of cleanup here to make that more obvious to the compiler (and subsequent developers)

IDisposable commented on src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -850,10 +850,10 @@ public override MemberInfo[] GetMembers(BindingFlags bindingAttr)
             if (types == null || types.Length == 0)
             {
                 // no arguments
+                PropertyInfo firstCandidate = candidates[0];
```
Hoisted for throw.

IDisposable commented on src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -894,7 +894,7 @@ public override MemberInfo[] GetMembers(BindingFlags bindingAttr)
                 {
                     if (match != null)
-                        throw new AmbiguousMatchException();
+                        ThrowHelper.ThrowAmbiguousMatchException(match);
```
`match` points at the first instance as usual

IDisposable commented on src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -923,7 +923,7 @@ public override MemberInfo[] GetMembers(BindingFlags bindingAttr)
                     {
                         if (ReferenceEquals(fieldInfo.DeclaringType, match.DeclaringType))
-                            throw new AmbiguousMatchException();
+                            ThrowHelper.ThrowAmbiguousMatchException(match);
```
and again... `match` points at the first instance as usual

IDisposable commented on src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -935,7 +935,7 @@ public override MemberInfo[] GetMembers(BindingFlags bindingAttr)
 
             if (multipleStaticFieldMatches && match!.DeclaringType!.IsInterface)
-                throw new AmbiguousMatchException();
+                ThrowHelper.ThrowAmbiguousMatchException(match);
```
...and again.

IDisposable commented on src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -987,7 +987,7 @@ public override MemberInfo[] GetMembers(BindingFlags bindingAttr)
                 {
                     if (match != null)
-                        throw new AmbiguousMatchException();
+                        ThrowHelper.ThrowAmbiguousMatchException(match);
```
... and again...

IDisposable commented on src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:
```diff
@@ -1015,7 +1015,7 @@ public override MemberInfo[] GetMembers(BindingFlags bindingAttr)
                 {
                     if (match != null)
-                        throw new AmbiguousMatchException();
+                        ThrowHelper.ThrowAmbiguousMatchException(match);
```
... and for the last time :(

IDisposable commented:
If this seems like a reasonable approach, I will build out tests for all the paths

@steveharter, @marek-safar, @MichalStrehovsky I am **not** sure I actually need to handle the case of a null `DeclaringType`... is that really a possibility, because much of the code I did a surface scan of seems to think that never happens, in which case I can simplify a bit and eliminate the `_Simple` strings.

IDisposable commented:
Check failures do not look to be related to anything I did,

IDisposable commented:
Fleshed out the rest of the AmbiguousMatchExeception uses (and removed talk about doing the same for AmbiguousImplementationExceptio as there's no details avaiable)

IDisposable reviewed (commented):
Cleanup and fleshed things out a bit more. Seems to build.

buyaa-n commented on src/coreclr/System.Private.CoreLib/src/System/Attribute.CoreCLR.cs:
```diff
@@ -436,6 +436,11 @@ private static AttributeUsageAttribute InternalGetAttributeUsage(Type type)
         private static Attribute[] CreateAttributeArrayHelper(Type elementType, int elementCount) =>
             elementType.ContainsGenericParameters ? new Attribute[elementCount] : (Attribute[])Array.CreateInstance(elementType, elementCount);
+
+        private static AmbiguousMatchException GetAmbiguousMatchException(Attribute attribute)
+        {
+            return new AmbiguousMatchException(SR.Format(SR.RFLCT_AmbigCust_Detailed, attribute.GetType().ToString(), attribute.ToString()));
+        }
```
NIT:
```suggestion
        private static AmbiguousMatchException GetAmbiguousMatchException(Attribute attribute) =>
            new AmbiguousMatchException(SR.Format(SR.RFLCT_AmbigCust_Detailed, attribute.GetType().ToString(), attribute.ToString()));
```

buyaa-n replied:
Actually why passing a same info twice? 

`attribute.GetType().ToString(), attribute.ToString()` produces `System.SerializableAttribute System.SerializableAttribute`

IDisposable replied:
Applied this fix and swapped in `SR.RFLCT_AmbigCust_Simple` for `SR.RFLCT_AmbigCust_Detailed`

IDisposable replied:
(and then simplified even more with rest of feedback)

buyaa-n commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -3299,8 +3305,8 @@
-  <data name="RFLCT_AmbigCust" xml:space="preserve">
-    <value>Multiple custom attributes of the same type found.</value>
+  <data name="RFLCT_AmbigCust_Detailed" xml:space="preserve">
+    <value>Multiple custom attributes of the same type found matching: '{0} {1}'.</value>
```
```suggestion
    <value>Multiple custom attributes of the same type found matching: '{0}'.</value>
```
Or if it looks better
```suggestion
    <value>Multiple custom attributes of the same type '{0}' found.</value>
```

IDisposable replied:
Switched to this message format.

buyaa-n commented on src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs:
```diff
@@ -575,6 +577,26 @@ internal static void ThrowFormatIndexOutOfRange()
+        internal static AmbiguousMatchException GetAmbiguousMatchException(CustomAttributeData customAttributeData)
+        {
+            Type attributeType = customAttributeData.GetType();
+            return new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Detailed, attributeType.ToString(), customAttributeData.ToString()));
```
`.ToString()` is redundant here and everywhere else

IDisposable replied:
Snipped.

buyaa-n commented on src/libraries/System.Private.CoreLib/src/System/DefaultBinder.cs:
```diff
@@ -383,8 +383,10 @@ internal partial class DefaultBinder : Binder
+
             if (ambig)
-                throw new AmbiguousMatchException();
+                throw ThrowHelper.GetAmbiguousMatchException(bestMatch);
```
NIT: the name `bestMatch` doesn't make much sense for throwing cases, could use something like `match` or if the name already used then something like `foundMatch`

buyaa-n replied:
Hm, I see it was already using this name in other places, never mind

buyaa-n commented on src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs:
```diff
@@ -575,6 +577,26 @@ internal static void ThrowFormatIndexOutOfRange()
+        internal static AmbiguousMatchException GetAmbiguousMatchException(Attribute attribute)
+        {
+            Type attributeType = attribute.GetType();
+            return new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Detailed, attributeType.ToString(), attribute.ToString()));
```
```suggestion
            return new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Detailed, attribute));
```

buyaa-n commented on src/libraries/System.Reflection.Context/src/System/Reflection/Context/ThrowHelper.cs:
```diff
@@ -0,0 +1,32 @@
+        }
+        internal static AmbiguousMatchException GetAmbiguousMatchException(CustomAttributeData customAttributeData)
+        {
+            Type attributeType = customAttributeData.GetType();
+            throw new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Detailed, attributeType.ToString(), customAttributeData.ToString()));
```
`customAttributeData.GetType().ToString()` produces `System.Reflection.RuntimeCustomAttributeData` not useful at all,  `customAttributeData.ToString()` prints the attribute type anyways.
```suggestion
            throw new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Detailed, customAttributeData));
```

IDisposable commented:
Thanks TONS for the review @buyaa-n ! I just had eye surgery so I'm face down with minimal screen time for at least a week, will integrate your changes ASAP!

IDisposable commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -3299,8 +3305,8 @@
   <data name="Resources_StreamNotValid" xml:space="preserve">
     <value>Stream is not a valid resource file.</value>
   </data>
-  <data name="RFLCT_AmbigCust" xml:space="preserve">
```
@buyaa-n  do you know the answer to this?

buyaa-n replied:
Not sure if I understood the question, `xml:space="preserve"` is for preserving the whitespaces in the string not the string. Even if the given string doesn't have whitespaces that need to be preserved it could be useful in case the message updated with a string that have a carriage return. For more info: https://learn.microsoft.com/en-us/dotnet/desktop/xaml-services/white-space-processing#white-space-in-inner-text-and-string-primitives

IDisposable replied:
My question was about me _removing_ a string from the resx file? Is that okay (since will not use this anymore) or should I just leave it in the resx as an orphan?

buyaa-n replied:
Ah sorry for misunderstanding, unused strings should be removed

IDisposable commented:
I merge the simple ones, but this is no an unstable revision until I loop through the rest... I'll resync with main and squash this all before making it ready again

IDisposable commented:
Clean build, ready for final re-review @buyaa-n @marek-safar @MichalStrehovsky

buyaa-n commented on src/coreclr/System.Private.CoreLib/src/System/Attribute.CoreCLR.cs:
```diff
@@ -436,6 +436,9 @@ private static AttributeUsageAttribute InternalGetAttributeUsage(Type type)
         private static Attribute[] CreateAttributeArrayHelper(Type elementType, int elementCount) =>
             elementType.ContainsGenericParameters ? new Attribute[elementCount] : (Attribute[])Array.CreateInstance(elementType, elementCount);
+
+        private static AmbiguousMatchException GetAmbiguousMatchException(Attribute attribute) =>
```
Why don't using the one in `ThrowHelper`? https://github.com/dotnet/runtime/blob/2887a9c4756783b81da2cabb8a5e2402409d0495/src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs#L612

IDisposable replied:
Good call.

buyaa-n commented on src/coreclr/System.Private.CoreLib/src/System/Attribute.CoreCLR.cs:
```diff
@@ -515,10 +518,11 @@ public static bool IsDefined(MemberInfo element, Type attributeType, bool inheri
+                return match;
 
-            throw new AmbiguousMatchException(SR.RFLCT_AmbigCust);
+            throw GetAmbiguousMatchException(match);
```
Here and all similar cases in this file
```suggestion
            throw ThrowHelper.GetAmbiguousMatchException(match);
```

buyaa-n commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -172,18 +172,27 @@
     <value>Ambiguous match found.</value>
   </data>
+  <data name="Arg_AmbiguousMatchException_Simple" xml:space="preserve">
+    <value>Ambiguous match for '{0}' found.</value>
```
NIT: This looks better to me
```suggestion
    <value>Ambiguous match found for '{0}'.</value>
```

IDisposable replied:
I agree... it used to be that way (with a `:`) but I was following a comment in a previous review... should I flip back?

buyaa-n commented on src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs:
```diff
@@ -599,6 +601,25 @@ internal static void ThrowFormatIndexOutOfRange()
+            return new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Attribute, attribute));
+        }
+
+        internal static AmbiguousMatchException GetAmbiguousMatchException(CustomAttributeData customAttributeData)
+        {
+            Type attributeType = customAttributeData.GetType();
+            return new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Detailed, attributeType, customAttributeData));
+        }
+
```
Is this used?
```suggestion
```

IDisposable replied:
Yes, it's in [src\coreclr\nativeaot\System.Private.CoreLib\src\System\Reflection\Attribute.NativeAot.cs](https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Attribute.NativeAot.cs#L124)

buyaa-n replied:
OK, then I don't think we need to pass the `customAttributeData.GetType()` into the message. A sample test for `customAttributeData.GetType().ToString()` produces `System.Reflection.RuntimeCustomAttributeData` which I believe will be the same in most attributes if not all.

buyaa-n commented on src/libraries/System.Reflection.Context/src/System/Reflection/Context/ThrowHelper.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    internal static class ThrowHelper
+    {
+        internal static AmbiguousMatchException GetAmbiguousMatchException(MemberInfo memberInfo)
```
Let's not create a new type for just one throw case, also [this project is dead](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Reflection.Context/README.md), probably no need to even update the exception message

IDisposable replied:
Would you prefer that my final push just leave out this entire assembly?

IDisposable replied:
Left it in, but inlined the code instead of creating a ThrowHelper

buyaa-n commented on src/libraries/System.Reflection.Context/src/System.Reflection.Context.csproj:
```diff
@@ -20,6 +20,7 @@ System.Reflection.Context.CustomReflectionContext</PackageDescription>
     <Compile Include="System\Reflection\Context\CollectionServices.cs" />
     <Compile Include="System\Reflection\Context\CustomReflectionContext.cs" />
     <Compile Include="System\Reflection\Context\CustomReflectionContext.Projector.cs" />
+    <Compile Include="System\Reflection\Context\ThrowHelper.cs" />
```
```suggestion
```

buyaa-n commented on src/libraries/System.Reflection.Context/src/System/Reflection/Context/Custom/CustomType.cs:
```diff
@@ -266,10 +265,11 @@ public override MethodInfo[] GetMethods(BindingFlags bindingAttr)
-                    throw new AmbiguousMatchException();
+                    return match;
+
+                throw ThrowHelper.GetAmbiguousMatchException(match);
```
```suggestion
                throw new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Simple, match));
```

buyaa-n commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -172,18 +172,27 @@
+    <value>Ambiguous match for '{0}' found.</value>
+  </data>
+  <data name="Arg_AmbiguousMatchException_Detailed" xml:space="preserve">
+    <value>Ambiguous match for '{0} {1}' found.</value>
```
I am still not sure we really need this version with `memberInfo.DeclaringType` added, for me just `Ambiguous match for '{0}' found.` with the only `memberInfo` passed is good enough.

IDisposable replied:
That's mostly helpful for the `AttributeData` case, but I can go either way :)

buyaa-n commented:
> Clean build, ready for final re-review

Thank you @IDisposable, final review done, and left few comments, overall LGTM.

Though I am not sure about having 2 versions of messages for throwing `AmbiguousMatchException`. Fore me just `Ambiguous match for '{0}' found.` version with the `memberInfo` passed is good enough. I am not sure how useful it will be when the  `memberInfo.DeclaringType` is added along with `memberInfo` into the exception message `Ambiguous match for '{0} {1}' found.`. Please let me know what you think, thanks

IDisposable commented:
Addressed final feedback, rebased and squashed.

buyaa-n commented on src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs:
```diff
@@ -605,6 +607,25 @@ internal static void ThrowFormatIndexOutOfRange()
+
+        internal static AmbiguousMatchException GetAmbiguousMatchException(CustomAttributeData customAttributeData)
+        {
+            Type attributeType = customAttributeData.GetType();
+            return new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Detailed, attributeType, customAttributeData));
```
```suggestion
            return new AmbiguousMatchException(SR.Format(SR.Arg_AmbiguousMatchException_Simple, customAttributeData));
```

buyaa-n commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/CustomAttributes/CustomAttributeHelpers.cs:
```diff
@@ -17,10 +17,10 @@ public static CustomAttributeNamedArgument ToCustomAttributeNamedArgument(this T
-                throw new AmbiguousMatchException();
-
-            return new CustomAttributeNamedArgument(members[0], new CustomAttributeTypedArgument(argumentType!, value));
+                throw ThrowHelper.GetAmbiguousMatchException(match);
+            return new CustomAttributeNamedArgument(match, new CustomAttributeTypedArgument(argumentType!, value));
```
```suggestion
                throw ThrowHelper.GetAmbiguousMatchException(match);
                
            return new CustomAttributeNamedArgument(match, new CustomAttributeTypedArgument(argumentType!, value));
```

IDisposable replied:
Just adding a blank line here? Cool

IDisposable commented:
Not sure why the Detailed version isn't useful, but if we're killing off the uses, I can remove the literal. I added these originally because what I was getting back from the various `.ToString()` calls **was different** information, so it seemed valuable.

buyaa-n commented:
> Not sure why the Detailed version isn't useful, but if we're killing off the uses, I can remove the literal. I added these originally because what I was getting back from the various `.ToString()` calls **was different** information, so it seemed valuable.

It is more about the extra conditional added for checking if the `declaringType` is not null, than it was not useful. At least throwing an exception is not a hot path that needs to be performant, so I am OK to add it if the declaringType info added is really useful, for which I am not that sure about. Let's take the example you used, for:
`Ambiguous implementation found for 'System.String Int32 Length'`
`Ambiguous implementation found for 'Int32 Length'`
How useful the `System.String` part for you, or is this could cause any confusion when you were actually looking for the `Length` property? 

Also, another option could be just passing the declaring type without null check as string.Format accepts null. The message would just have and extra space when the declaring type is null `Ambiguous implementation found for ' Int32 Length'`. How this sound?

buyaa-n commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/CustomAttributes/CustomAttributeHelpers.cs:
```diff
@@ -17,10 +17,11 @@ public static CustomAttributeNamedArgument ToCustomAttributeNamedArgument(this T
-
-            return new CustomAttributeNamedArgument(members[0], new CustomAttributeTypedArgument(argumentType!, value));
+                throw ThrowHelper.GetAmbiguousMatchException(match);
+                
```
```suggestion

```

IDisposable commented:
> Let's take the example you used, for: Ambiguous implementation found for 'System.String Int32 Length'
> Ambiguous implementation found for 'Int32 Length'
>
> How useful the System.String part for you, or is this could cause any confusion when you were actually looking for the Length property?

When we're figuring out why the exception was thrown (at post-mortem debug), knowing the exact class that had ONE of the ambiguities lets us narrow things down. We wouldn't be doing this on core classes usually, so having the type seems very helpful for opening the right source file :)

> Also, another option could be just passing the declaring type without null check as string.Format accepts null. The message would just have and extra space when the declaring type is null Ambiguous implementation found for ' Int32 Length'. How this sound?

I like that idea... lemme churn that a bit and commit.

IDisposable committed:
Add message for AmbiguousMatchException

Add information to what is ambiguously matched when an AmbiguousMatchException is thrown as we know what we found multiple matching items for.

Addressed review feedback 
Renames of SR/RESX strings. Removal of `.ToString()` calls. Remove unused strings from RESX RESX string rewording.
Change the exception format string to "found for" style.
Cleanup of DefaultBinder.cs FindMostDerived* to use the local variable.
Remove copy-pasta ThrowHelper.cs instance.
Don't emit the CustomAtrributeData's type
All MemberInfo cases pass a (possibly null) `DeclaringType`.
Co-authored-by: Buyaa Namnan <buyankhishig.namnan@microsoft.com>

IDisposable commented:
Incorporated these last few discussions. Now always passes the (possibly null) `DeclaringType` when dumping a MemberInfo (or derived). The other ResX messages are now named according to the ambiguous thing, so we're now talking about _AmbiguousMatchException_Assembly_, _Arg_AmbiguousMatchException_Attribute_, _Arg_AmbiguousMatchException_CustomAttributeData_ (was Simple), _Arg_AmbiguousMatchException_MemberInfo_ (was Detailed),  and _Arg_AmbiguousMatchException_RoDefinitionType_ with matching ThrowHelper or inline versions (as needed)

IDisposable commented:
Build errors, please hold

buyaa-n reviewed (approved):
Thank you @IDisposable LGTM

buyaa-n commented:
The failures unrelated and all known issues, merging
